lib/CapabilityInvocation.js

```diff
@@ -0,0 +1,90 @@
+ *
+ * Neither the name of the Digital Bazaar, Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+ * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+'use strict';
+
+const jsigs = require('jsonld-signatures');
+const utils = require('./utils');
+const vocab = require('./vocab');
+
+const {ProofPurpose} = jsigs;
+
+module.exports = class CapabilityInvocation extends ProofPurpose {
+  constructor(injector) {
+    super(injector);
+    this.documentLoader = injector.use('jsonld').documentLoader;
+    this.uri = vocab.capabilityInvocationUri;
+  }
+
+  // Arguments:
+  //  - document: An already-expanded version of the invocation document
+  //    this proof is attached to
+  //  - proof: An already-expanded version of the proof we are checking
+  //
+  // proofPurposeOptions keywords:
+  //  - expectedTarget: the target we expect this capability to apply to (URI).
+  //  - caveatVerifiers: a hashmap of caveat URIs to procedures which
+  //    verify those caveats.  The caveat-verifying procedures should take
+  //    two arguments: the caveat being verified, and the expanded invocation
+  //    document.
+  async verify(document, proof, proofPurposeOptions) {
+    const {capabilityUri, creatorUri} = vocab;
+
+    // Retrieve the JSON-LD document associated with the capability ID
+    const {document: capability} = await this.documentLoader(
+      utils.getOneOrDie(proof[capabilityUri]));
+
+    // Prepare to check that the creator matches the parentCapability's invoker
+    const creator = utils.getOneOrDie(proof[creatorUri]);
+    const parentInvoker = await utils.getParentInvoker({
+      capability: parent,
+      documentLoader :this.documentLoader
+    });
+
+    const verifiedCaveats = utils.verifyCaveats({
+      capability: document,
+      proofPurposeOptions
+    });
+    const verifiedCapability = utils.verifyCap({
+      capability,
+      jsigs: this.injector.use('jsonld-signatures'),
+      proofPurposeOptions
+    });
+    return (parentInvoker === creator && // proof created by authorized invoker
+            await verifiedCaveats &&
+            await verifiedCapability);
```

Comment:
I see you separated these out into variables above as opposed to putting them between the && as I originally did it.  That's fine, though it's a little bit less optimal because it doesn't "short circuit early" as the other code did... it may run checks unnecessarily once we find out the previous checks don't pass anyway.

Reply:
I see, yes you're right!  I actually didn't realize you could use await that way.